### PR TITLE
[Lora] Seperate logic

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR separates old and new (PEFT) lora logic better to make it easier to remove the old LoRA logic.

### CODE REVIEW

File: src/diffusers/models/lora.py [hunk 1]
```diff
@@ -57,25 +66,6 @@ def text_encoder_mlp_modules(text_encoder):
     return mlp_modules
 
 
-def text_encoder_lora_state_dict(text_encoder):
```
Comment (patrickvonplaten) [inline src/diffusers/models/lora.py:60]: This is only used in training scripts, let's remove it from here. It'll make it easier to later completely remove once the training scripts are refactored.

File: src/diffusers/loaders/lora.py [hunk 2]
```diff
@@ -66,6 +67,35 @@
 LORA_DEPRECATION_MESSAGE = "You are using an old version of LoRA backend. This will be deprecated in the next releases in favor of PEFT make sure to install the latest PEFT and transformers packages in the future."
 
 
+def text_encoder_attn_modules(text_encoder):
```
Comment (patrickvonplaten) [inline src/diffusers/loaders/lora.py:70]: We need those for PEFT for now, so to be able to easily delete `src/diffusers/models/lora.py` let's move it here.

File: examples/text_to_image/train_text_to_image_lora.py [hunk 3]
```diff
@@ -41,7 +41,7 @@
 import diffusers
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:41]: Missing example refactor from this PR: https://github.com/huggingface/diffusers/pull/5331

File: examples/dreambooth/train_dreambooth_lora.py [hunk 4]
```diff
@@ -70,6 +70,38 @@
 logger = get_logger(__name__)
 
 
+def text_encoder_lora_state_dict(text_encoder):
+    state_dict = {}
+
+    def text_encoder_attn_modules(text_encoder):
+        from transformers import CLIPTextModel, CLIPTextModelWithProjection
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:77]: Why is this needed? We can directly import it at the top.

File: examples/dreambooth/train_dreambooth_lora.py [hunk 5]
```diff
@@ -70,6 +70,38 @@
 logger = get_logger(__name__)
 
 
+def text_encoder_lora_state_dict(text_encoder):
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:73]: Also, I see that this exact same function is copied over 4 times. So, let's move it to `training_utils.py` please.

File: src/diffusers/loaders/lora.py [hunk 6]
```diff
@@ -1339,7 +1368,7 @@ def process_weights(adapter_names, weights):
             )
         set_weights_and_activate_adapters(text_encoder, adapter_names, text_encoder_weights)
 
-    def disable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"] = None):
+    def disable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"] = None):  # noqa: F821
```
Comment (sayakpaul) [inline src/diffusers/loaders/lora.py:1371]: Why the noqa?

**Review by sayakpaul — APPROVED** [review]
All in for sweeping off 🧹

Thanks!

File: examples/dreambooth/train_dreambooth_lora.py (hunk 5 above)
Comment (BenjaminBossan) [reply examples/dreambooth/train_dreambooth_lora.py:73]: I think the idea to move it here was because the original function is being deprecated. So I assume this function will be removed eventually. If that's so, do we want to add a `# TODO remove in diffusers vX.Y`? Not sure if there is such a convention in diffusers.

File: examples/text_to_image/train_text_to_image_lora.py [hunk 7]
```diff
@@ -458,25 +489,43 @@ def main():
...
-        lora_attn_procs[name] = LoRAAttnProcessor(
-            hidden_size=hidden_size,
-            cross_attention_dim=cross_attention_dim,
-            rank=args.rank,
+    unet_lora_parameters = []
+    for attn_processor_name, attn_processor in unet.attn_processors.items():
+        # Parse the attention module.
+        attn_module = unet
+        for n in attn_processor_name.split(".")[:-1]:
+            attn_module = getattr(attn_module, n)
```
Comment (BenjaminBossan) [inline examples/text_to_image/train_text_to_image_lora.py:497]: `operator.attrgetter` can be used to replace this pattern, but looping explicitly is also fine.

**Review by yiyixuxu — APPROVED** [review]
thanks!

File: examples/dreambooth/train_dreambooth_lora.py (hunk 4 above)
Comment (patrickvonplaten) [reply examples/dreambooth/train_dreambooth_lora.py:77]: So that we can remove `text_encoder_attn_modules` from src/diffusers

File: examples/dreambooth/train_dreambooth_lora.py (hunk 5 above)
Comment (patrickvonplaten) [reply examples/dreambooth/train_dreambooth_lora.py:73]: Yeah good idea to add:
`# TODO remove ...` statements

@sayakpaul we should try to remove this function once training scripts are changed to use PEFT, I don't want to move them back into src/diffusers

File: src/diffusers/loaders/lora.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/diffusers/loaders/lora.py:1371]: Some linters complain about `PreTrainedModel` not being preset
